### DISCUSSION TITLE
NEED HELP : Reading data from a flowmeter on a microcontroller

### DISCUSSION THREAD

**Discussion by MMetereau** [community/community#62612] [Programming Help]
### Body

For a personal DIY project, I want to make an app that measure water consumption and send the data to an API.

I have the Keyence FD-Q50C clamp on flowmeter that has a screen to display the data. The flowmeter can be connected to the IO Link Master module Keyence NQ-MP8L which has a M12 to Ethernet cable that can transmit the data. Keyence offers the NQ Sensor Monitor software to configure the NQ-MP8L, by linking the module to my PC with the Ethernet cable I managed to configure the IP of the module.

Now I would like to read the data on a microcontroller. On the NQ-MP8L datasheet, it is said that the module is compatible with different data sending protocol: Ethernet/IP, Modbus, Profinet.

I first wanted to see the data format so I used Wireshark and I observed that I was receiving some CIP packets from the module (I don't have a pipe on which I can plug the flowmeter so I don't know if it would send some different data with a flow or if it send data even when the flow is null).

I have now some difficulties to understand how the data are transmitted and their format in order to read them on a microcontroller. I have a microntroller (W5500 EVB Pico) with an Etheret Port on which I can plug the NQ-MP8L but I am open to suggestions.

Here are the datasheets : https://drive.google.com/drive/folders/1tGxoTztU6-aOUp8EyHqI-dLpJ0zSU3lD?usp=sharing (I also added the datasheet of the FD-Q50C just in case)

If someone has any idea of how I can do read those data I would be glad to hear it.

### Guidelines

- [X] I have read and understood this category's [guidelines](https://github.com/orgs/community/discussions/51384) before making this post.

**Comment by shivvamm** [comment]
Creating an app to measure water consumption using the Keyence FD-Q50C clamp on flowmeter and sending the data to an API sounds like an interesting DIY project! Here's a high-level overview of the steps you might need to take:

1. **Data Format and Protocol Understanding:**
   - Since you've observed CIP packets using Wireshark, it's important to understand the Common Industrial Protocol (CIP) used by the Ethernet/IP protocol. Refer to the Ethernet/IP specification to decode the CIP packets and extract the relevant data.
   - You might also need to understand the flow data structure and how it's encapsulated within the CIP packets. The datasheets you provided should help you decipher this information.

2. **Microcontroller Communication:**
   - The W5500 EVB Pico microcontroller with Ethernet capability is a suitable choice. It supports the Ethernet protocol, which is compatible with the communication options provided by the NQ-MP8L module (Ethernet/IP, Modbus, Profinet).
   - You'll need to implement the chosen communication protocol on your microcontroller to establish communication with the NQ-MP8L module.

3. **Protocol Implementation:**
   - Depending on the chosen protocol (Ethernet/IP, Modbus, Profinet), you'll need to implement the corresponding communication protocol on your microcontroller.
   - This might involve socket programming, packet parsing, and data extraction based on the protocol specifications. Libraries or examples for these protocols might be available for your microcontroller platform.

4. **Integration and Data Parsing:**
   - Once you establish communication and start receiving data from the NQ-MP8L module, you'll need to parse the incoming data according to the structure you identified earlier.
   - Extract the relevant flow data from the received packets.

5. **API Integration:**
   - Implement the logic to send the extracted flow data to your desired API. You might need to format the data according to the API's requirements (JSON, XML, etc.).
   - The W5500 EVB Pico might have HTTP or TCP/IP libraries that you can use to make API requests.

6. **User Interface (Optional):**
   - If you want a user interface, you can create a simple app that displays the water consumption data received from the microcontroller.
   - Depending on your skills and preferences, you can develop a desktop or mobile app. Many platforms and frameworks are available for this purpose.

Remember that working with industrial protocols and hardware might be complex and require a good understanding of networking, protocols, and microcontroller programming. It's a great learning experience, but it might involve some challenges along the way.

As you work on this project, don't hesitate to consult relevant documentation, seek help from online communities, and iterate on your approach as needed. Good luck, and have fun with your DIY endeavor!

**Comment by Spixmacaw** [comment]
responce #1
